essentia: init at 2.1_beta5

doronbehar opened the pull request:
###### Motivation for this change

Can help bring support for the [beets](http://beets.io/) plugin [AcousticBrainz Submit Plugin](https://beets.readthedocs.io/en/v1.4.9/plugins/absubmit.html#acousticbrainz-submit-plugin).

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

doronbehar committed:
essentia: init at 2.1_beta5

doronbehar commented:
Depends upon #70372

aszlig commented on pkgs/development/libraries/essentia/default.nix:
```diff
@@ -0,0 +1,83 @@
+    ]
+    ++ lib.optionals (gaiaSupport) [ gaia ]
+  ;
+  wafConfigureFlags = [
```
Superfluous empty list.

doronbehar replied:
Same as in https://github.com/NixOS/nixpkgs/pull/70372#discussion_r331733095 .

aszlig commented on pkgs/development/libraries/essentia/default.nix:
```diff
@@ -0,0 +1,83 @@
+  ;
+  wafConfigureFlags = [
+  ]
+    ++ lib.optionals (pythonSupport) [ "--with-python" ]
```
Note that you can also use `lib.optional` for this and most of the others, which implies a singleton list, eg.:

```suggestion
    ++ lib.optional (pythonSupport) "--with-python"
```

doronbehar replied:
Same as in https://github.com/NixOS/nixpkgs/pull/70372#discussion_r331733772 .

aszlig commented on pkgs/top-level/python-packages.nix:
```diff
@@ -5008,6 +5008,11 @@ in {
   transitions = callPackage ../development/python-modules/transitions { };
 
+  essentia = (toPythonModule (pkgs.essentia.override {
+    pythonPackages = python;
```
Did you mean:

```suggestion
    pythonPackages = self;
```

... here?

aszlig replied:
Also, since the package set is for a particular version, you'd also need to pass `python` here as well.

doronbehar replied:
Please tell me what do you think of the last changes I pushed.

aszlig commented on pkgs/development/libraries/essentia/default.nix:
```diff
@@ -0,0 +1,83 @@
+  };
+  patches = [
+    (fetchpatch {
+      url = "https://github.com/MTG/essentia/pull/915.patch";
```
Please make sure to add a comment (eg. `Make waf use ${PREFIX} to install examples` - the title of your pull request) about what this patch does because it's not clear from the filename.

doronbehar replied:
:+1:

aszlig commented on pkgs/development/libraries/essentia/default.nix:
```diff
@@ -0,0 +1,83 @@
+
+  meta = with lib; {
+    homepage = "https://github.com/MTG/essentia";
+    description = "C++ library with python bindings which implements similarity measures and classiﬁcations on the results of audio analysis, and generates classiﬁcation models that Essentia can use to compute high-level description of music";
```
Similar to https://github.com/NixOS/nixpkgs/pull/70372#discussion_r331700776:

```suggestion
    description = "Library for audio and music analysis, description and synthesis";
```

doronbehar replied:
:+1:.

aszlig commented on pkgs/development/libraries/essentia/default.nix:
```diff
@@ -0,0 +1,83 @@
+# Probably because gaia which depends on qt4
+, qt4
+, wafHook
+, pythonSupport ? true
```
Given that you also expose this via `pythonPackages`, doesn't it make sense to set this to `false` instead, so that Python support is left out when not explicitly using `pythonPackages.essentia`?

doronbehar replied:
You are right, I totally agree.

aszlig commented on pkgs/development/libraries/essentia/default.nix:
```diff
@@ -0,0 +1,83 @@
+    ++ lib.optionals (withExamples) [ "--with-examples" ]
+  ;
+
+  meta = with lib; {
```
Note that it might also be a good idea to add a [`platforms` attribute](https://nixos.org/nixpkgs/manual/#sec-standard-meta-attributes) here, because the change in https://github.com/NixOS/nixpkgs/pull/70392 doesn't enable the `absubmit` plugin by default and thus once someone is enabling the plugin, they won't get binaries for `essentia` from the cache.

doronbehar replied:
Right, thanks for catching that.

doronbehar reviewed (commented):
Commented and pushed changes.

aszlig commented on pkgs/development/libraries/essentia/default.nix:
```diff
@@ -40,6 +41,8 @@ stdenv.mkDerivation rec {
   };
   patches = [
     (fetchpatch {
+      # Make waf use ${PREFIX} to install examples - otherwise, our build fail
```
```suggestion
      # Make waf use ${PREFIX} to install examples - otherwise, our build fails
```

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review-may-2019/3032/71

matthewbauer commented:
@grahamcofborg eval

doronbehar commented:
It seems the evaluation has failed because of https://github.com/NixOS/nixpkgs/commit/a3a782ec5f7c5515cb0faecf4c09d0eb14888f50 I wonder why I haven't chose it in the first place.

Rebased to master and fixed evaluation.

jonringer commented on pkgs/top-level/python-packages.nix:
```diff
@@ -5153,6 +5153,13 @@ in {
   transitions = callPackage ../development/python-modules/transitions { };
 
+  essentia = (toPythonModule (pkgs.essentia.override {
+    pythonPackages = self;
```
instead of passing the package scope, you should pass the interpreter, then pythonPackages -> python.pkgs
```suggestion
    inherit python;
```

doronbehar replied:
Inheriting `python` works however. Marking this as resolved, yet not committing & pushing.

jonringer commented on pkgs/top-level/python-packages.nix:
```diff
@@ -5153,6 +5153,13 @@ in {
+    pythonPackages = self;
+    pythonSupport = true;
+    # Use our gaia
+    gaia = self.gaia;
```
this could be combined with the other suggestion
```suggestion
    inherit python gaia;
```

doronbehar replied:
From some reason I get this error when inheriting `gaia`:

```
error: undefined variable 'gaia' at /var/src/nixpkgs/pkgs/top-level/python-packages.nix:5156:55
```

Where that line is where `essentia` is defined.

jonringer replied:
interesting, well, callPackage should be able to pick up gaia from the current scope. So It might not be needed at all

doronbehar replied:
Makes sense, I've pushed a fix but for `python3Packages.essentia` we depend on https://github.com/NixOS/nixpkgs/pull/69310 .

doronbehar commented:
BTW Thanks for commenting @jonringer !

jonringer reviewed (commented):
I would say squash this into 2 commits:

```
python3Packages.gaia: enable for python3
essentia: init at 2.1_beta5
```

doronbehar committed:
pythonPackages.gaia: Enable for Python3

doronbehar commented:
Closing. There's a mess with the commits here + there are too much build errors upstream fixed but haven't made a release yet. I'll open a new PR when there's a new release upstream.
